src/main/java/org/benetech/servicenet/filters/RequestFilter.java

```diff
@@ -0,0 +1,120 @@
+package org.benetech.servicenet.filters;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import java.util.Objects;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.benetech.servicenet.domain.RequestLogger;
+import org.benetech.servicenet.service.RequestLoggerService;
+import org.benetech.servicenet.service.mapper.RequestLoggerMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.Ordered;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+import org.springframework.web.util.ContentCachingRequestWrapper;
+import org.springframework.web.util.ContentCachingResponseWrapper;
+import org.springframework.web.util.WebUtils;
+
+@Component
+public class RequestFilter extends OncePerRequestFilter implements Ordered {
+
+    @Autowired
+    private RequestLoggerService requestLoggerService;
+
+    @Autowired
+    private RequestLoggerMapper requestLoggerMapper;
+
+    @Override
+    public int getOrder() {
+        return Ordered.LOWEST_PRECEDENCE - 8;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+        throws ServletException, IOException {
+        ContentCachingRequestWrapper wrappedRequest = new ContentCachingRequestWrapper(request);
+        ContentCachingResponseWrapper wrappedResponse = new ContentCachingResponseWrapper(response);
+
+        filterChain.doFilter(wrappedRequest, wrappedResponse);
+
+        Map<String, Object> trace = getTrace(wrappedRequest);
+        Map<String, Object> responseTrace = getResponseTrace(wrappedResponse);
+
+        getBody(wrappedRequest, trace);
+        getResponseBody(wrappedResponse, responseTrace);
+
+        RequestLogger requestLogger = new RequestLogger();
+        requestLogger.setRemoteAddr(Objects.toString(trace.get("remoteAddr"), ""));
+        requestLogger.setRequestUri(Objects.toString(trace.get("requestUri"), ""));
+        requestLogger.setRequestMethod(Objects.toString(trace.get("requestMethod"), ""));
+        requestLogger.setRequestParameters(Objects.toString(trace.get("requestParams"), ""));
+        requestLogger.setRequestBody(Objects.toString(trace.get("body"), ""));
+        requestLogger.setResponseStatus(Objects.toString(responseTrace.get("status"), ""));
+        requestLogger.setResponseBody(Objects.toString(responseTrace.get("body"), ""));
+
+        requestLoggerService.save(requestLoggerMapper.toDto(requestLogger));
```

Comment:
I think it's worth measuring if it slows down the requests - maybe we could save them in a separate thread

Reply:
I do not spot any big difference